src/biogeochem/CNPhenologyMod.F90

```diff
@@ -2195,16 +2368,15 @@ subroutine CropPhase(bounds, num_pcropp, filter_pcropp, &
        p = filter_pcropp(fp)
 
        if (croplive(p)) then
-          ! Start with cphase_planted, but this might get changed in the later
-          ! conditional blocks.
-          crop_phase(p) = cphase_planted
           if (leafout(p) >= huileaf(p) .and. hui(p) < huigrain(p)) then
              crop_phase(p) = cphase_leafemerge
-          else if (hui(p) >= huigrain(p)) then
+          else if (hui(p) >= huigrain(p) .and. crop_inst%cphase_patch(p) >= cphase_leafemerge) then
```

Comment:
- [ ] I'm a little nervous about this change in subroutine CropPhase. There is some trickiness in driver ordering which was part of my motivation for introducing this CropPhase subroutine rather than just using the cphase variable: This CropPhase routine is called from the nutrient competition code. Annoyingly, CropPhenology is called *before* the nutrient competition stuff when using FUN, but *after* the nutrient competition stuff when not using FUN. I haven't completely thought through the implications of this trickiness for referencing crop_inst%cphase_patch (which is set in CropPhenology) here, but I'm thinking that the behavior will be wrong in the case where CropPhase is called after CropPhenology (because cphase_patch will have already been updated in this time step). One way I can see to fix this would be: At the start of the time step, save the current cphase_patch to a new variable like cphase_start_of_timestep_patch. Then I think that this conditional should be changed to look at cphase_start_of_timestep_patch instead of cphase_patch (and probably also change the similar conditional in CropPhenology: it shouldn't matter there but it would be clearer to keep them consistent).

Reply:
@samsrabin - I'm going to address your recent email here to keep everything in one place. You asked:

> One of the last things I need to do for the prescribed crop calendars PR is to potentially rework my changes to CropPhase(). Specifically, you suggested that I might need to do a little tweaking to handle the fact that CropPhase() can get called at different points in a timestep. I'm wondering what you think about what I wrote in my reply [here](https://github.com/ESCOMP/CTSM/pull/1863#discussion_r1066433795)—do you still think I need to add some special handling? And would your answer change if I can show that some test(s) is (are) b4b identical with my change?

I'll be honest: I don't remember the details of my thinking from my comment in October, but I have gotten my head back into this so I *think* I know what my concern was, and I think I *do* still have this concern. And unfortunately, this is a condition that I wouldn't trust the test suite to pick up. My understanding is that you added the extra condition (`crop_inst%cphase_patch(p) >= cphase_leafemerge`) here to stay consistent with your changes in `CropPhenology`, and that this change is designed to ensure that we stay in the vegetative phase for at least one time step.

A first potential problem that I'm seeing now, which I guess I didn't see before (though it may be that I'm seeing things wrong now) is that it looks like your logic in CropPhenology might not accomplish the goal I think you're trying to accomplish: I thought the intent was to ensure that the crop spends at least one time step in the vegetative phase (i.e., cphase_leafemerge) before going on to the reproductive phase. But what I'm seeing from the new logic in CropPhenology is that, in the case where hui >= huigrain and cphase < cphase_leafemerge, it looks like we'll set cphase = cphase_planted. Let me know if you think I'm reading this wrong - or if I'm misunderstanding the intent.

The rest of this comment will assume that you resolve the above, so that in the case where hui >= huigrain and cphase < cphase_leafemerge, we'll enter cphase_leafemerge for one time step... given that:

The problem I see then is: If we start the time step with cphase_patch < cphase_leafemerge, then in this time step hui comes to exceed huigrain, then we'll have: In the call to `CropPhenology`, we'll prevent cphase_patch from reaching cphase_grainfill, and according to what I think is meant, we'll set cphase_patch = cphase_leafemerge. But then, when CropPhase is called later in the time step, it will see a cphase_patch that is already cphase_leafemerge, so the conditional you added here (`crop_inst%cphase_patch(p) >= cphase_leafemerge`) will be true, leading to CropPhase returning a value of cphase_grainfill. Thus, CropPhase's return value would be inconsistent with cphase_patch in this time step.

So the way I see things is that the added conditional here isn't actually achieving what you want it to achieve (ensuring that the return value of CropPhase stays in the vegetative phase for at least one time step) and creates a possibility that, for this one time step, CropPhase will say that we're in grainfill when other parts of the code (which reference cphase_patch) still say we're in the vegetative stage. Whether this is a significant problem is a different question... my gut sense is that it probably is *not* (though I haven't reviewed all of the relevant code). But if it isn't a significant problem, then I would argue for removing your new condition (`crop_inst%cphase_patch(p) >= cphase_leafemerge`) from `CropPhase` (since I don't think it works as intended) and just adding a comment here acknowledging the difference between the calculation of cphase_patch and the operation of CropPhase in this respect (i.e., documenting that it's intentional that the two blocks of code don't completely match). Or you could go with my earlier suggestion of introducing `cphase_start_of_timestep_patch`.

(A way to expose this problem would be to design a scenario where this extra condition is needed – i.e., where hui jumps to being greater than huigrain before the crop has spent even a single time step in the vegetative phase – and then running this in a case with FUN. It's not that I expect unintended answer changes from this problem, but rather that I think you would *not* get answer changes in the nutrient competition code for this specific time step when your intent *is* to get answer changes in this situation. I don't actually think that it's worth the effort to design, run and analyze this scenario, but I just wanted to point out what it would require in theory.)

You raised the issue of hui also being updated in CropPhenology, and so possibly having the same issues... I haven't had a chance to think that through enough to respond intelligently.

Reply:
Just to put in writing a couple of things that I just talked with @samsrabin about:

- [ ] It *does* seem like we may need to change CropPhenology to achieve the desired behavior. @samsrabin will look more into that.
- @samsrabin will then think about how to change CropPhase to ensure consistency - or if we should just not try to keep things consistent for this one time step in the vegetative phase in this edge case scenario
- I think @samsrabin is right that we can be out of sync for a time step in the situation where hui is updated in CropPhenology, if CropPhase is called before CropPhenology. I have a feeling that I wasn't concerned about this when I implemented CropPhase because my biggest motivation at the time was maintaining bit-for-bit results with the previous code, and I'm guessing we already had the potential for being out-of-sync in the previous code, so I was just maintaining previous behavior. But if we wanted to ensure things are consistent moving forward, and don't mind the potential for minor answer changes, we could consider moving this hui update to wherever hui is generally set (earlier in the time step, presumably).

Reply:
It looks like I made these changes because, in the weird setup required for prescribed crop calendar runs, I was getting a floating-point error due to NaN `astemi` in the last line here:

https://github.com/ESCOMP/CTSM/blob/0b98af47bcd307ca04ceaa0ed03c254d6be0344e/src/biogeochem/CNAllocationMod.F90#L366-L369

`astemi` is only even set just above, in the "if crop is in vegetative phase" block:
https://github.com/ESCOMP/CTSM/blob/0b98af47bcd307ca04ceaa0ed03c254d6be0344e/src/biogeochem/CNAllocationMod.F90#L358

Ensuring that the crop went through the vegetative phase for at least one timestep thus ensured that `astemi` wouldn't be NaN. However, as discussed, this "solution" might cause other issues.

Instead, I should more directly ensure that `astemi` is never NaN at line 369. As you can see at line 358, `astemi` is just a place to save the value of `astem` at the end of the vegetative phase. Generalizing to include runs where the vegetative phase is skipped, we could say instead that it's a place to save the value of `astem` _at the beginning of the reproductive phase_. What I could do, then, just before line 369 in that reproductive phase block, is check whether `astemi` is NaN—if so, assigning it `astem`'s value.

That raises the question: Can I rely on `astem` being non-NaN at that point?

Later in this same subroutine, you can see `astem` is set to zero in the "planted but not yet vegetative" phase:
https://github.com/ESCOMP/CTSM/blob/0b98af47bcd307ca04ceaa0ed03c254d6be0344e/src/biogeochem/CNAllocationMod.F90#L396-L405

This makes me think I should just set `astem` (and all those values, actually) in PlantCrop(). Might as well set `astemi` too (also to zero), rather than doing the "check if `astemi` NaN, if yes, assign `astem`'s value" idea I floated above. 

I would do the same for `aleafi` to avoid just hitting a floating-point error at line 379:
https://github.com/ESCOMP/CTSM/blob/0b98af47bcd307ca04ceaa0ed03c254d6be0344e/src/biogeochem/CNAllocationMod.F90#L379

If I'm understanding correctly, this should do what I really wanted, allowing me to revert the changes to crop phase code I made in 7761d1af65.

What do you think?

Reply:
More simply than moving that pre-emergence block to PlantCrop(), I could just add the following after line 401:
```fortran
aleafi(p) = aleaf(p)
astemi(p) = astem(p)
```

Reply:
I haven't completely parsed all of your suggestions in the big comment, but I like the idea of addressing this issue more directly by ensuring that aleafi and astemi have reasonable values even if we skip straight to the grainfill stage. If you have a sense of the best way to accomplish this, I'm happy to go with your suggestion – maybe what you suggested in your small follow-on comment?

I think that checking for whether a variable is NaN would be fragile - especially in this case where it would become non-NaN in a given year, but then you'd (presumably) want it to return to NaN at the start of the next year. In principle this could work but we don't typically reassign values to NaN after initialization. That said, you could do something effectively the same by setting it to spval and then checking if it's spval. But it sounds like you may have a better solution in mind that doesn't rely on anything like that.

Let me know if you'd like to talk this through any further.

Reply:
Thanks, Bill. Once I get my test set up I'll try the simple solution to make sure that does what I want without messing anything else up.

Reply:
Okay, I have ripped out my changes to how `cphase` is updated in CropPhenology() to avoid conflict with how CropPhase() does it. Instead, I have added:
- Initializations of `aleafi` and `astemi` to 1 and 0, respectively, in the calc_crop_allocation_fractions() "pre emergence" and "not croplive" blocks.
- Initializations of those (along with `aleaf`, `astem`, `aroot`, and `arepr`) in PlantCrop().
This belt-and-suspenders approach ensures that `aleafi` and `astemi` can never be uninitialized when calc_crop_allocation_fractions() attempts to access them, which is what was causing the crash.

Unfortunately, I have been unable to reproduce the crash, even using the same setup I have in my notes from the initial development. Let me know if you'd prefer me to just rip out all these changes, keeping them in the back of my head in case I run into the crash again.

Reply:
Thanks a lot for all of your work on this, @samsrabin ! I don't have time to get my head fully into this right now, but I'm happy to go with your new set of changes, which follows what we discussed last month, if you feel pretty confident that it's doing the right thing now, even if you haven't been able to reproduce the crash.

Reply:
CropPhase() change resolved with 5c80d6b.